### PR TITLE
fix flaky file discovery tests

### PR DESCRIPTION
**Pull request by krasi-georgiev**
Fixes: https://github.com/prometheus/prometheus/issues/3419
fsnotify send invalid updates causing the fd service to send empty `group[]` struct before the test files are written to disk

Now using an extra sync channel to ignore the fd updates until the test files are written to disk.

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
cc @grobie

File: discovery/file/file_test.go [hunk 1]
```diff
@@ -59,30 +61,45 @@ func testFileSD(t *testing.T, prefix, ext string, expect bool) {
...
+	// To avoid empty group struct sent from the discovery caused by invalid fsnotify updates,
+	// drain the channel until we are ready with the test files.
+	filesReady := make(chan struct{})
+	go func() {
+	Loop:
+		for {
+			select {
+			case <-ch:
+			case <-filesReady:
+				break Loop
```
Comment (juliusv) [inline discovery/file/file_test.go]: Can't this just be `return`, since it's leaving the function afterward anyway?

Comment (krasi-georgiev) [reply discovery/file/file_test.go]: yeah.
it is failing again with travis so need to find another way.

**Comment by krasi-georgiev** [comment]
@grobie I think this solves it once and for all.

File: discovery/file/file_test.go [hunk 2]
```diff
@@ -27,11 +27,13 @@ import (
...
 )
 
+const testDir = "fixtures"
+
 func TestFileSD(t *testing.T) {
-	defer os.Remove("fixtures/_test_valid.yml")
-	defer os.Remove("fixtures/_test_valid.json")
-	defer os.Remove("fixtures/_test_invalid_nil.json")
-	defer os.Remove("fixtures/_test_invalid_nil.yml")
+	defer os.Remove(testDir + "/_test_valid.yml")
```
Comment (grobie) [inline discovery/file/file_test.go]: How does go on Windows work? Is the slash file separator supported? Or should we use Filepath.Join?

Comment (krasi-georgiev) [reply discovery/file/file_test.go]: yeah seems broken from the first commit , just proves it was never run on windows ;)

but anyway good catch. I made the changes.

File: discovery/file/file_test.go [hunk 3]
```diff
@@ -59,30 +61,45 @@ func testFileSD(t *testing.T, prefix, ext string, expect bool) {
...
 
-	timeout := time.After(15 * time.Second)
 	// The files contain two target groups.
+	// ready to receive discovery group updates so stop draining the channel
+	close(filesReady)
 retry:
 	for {
 		select {
-		case <-timeout:
+		case <-time.After(15 * time.Second):
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply discovery/file/file_test.go]: maybe I don't see anything obvious, but I don't think my change does anything but removing the `timeout ` variable and moving the timeout directly in the loop.

It should work as before. if no update is received form the file discovery service within 15 sec than fail the test.

[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply discovery/file/file_test.go]: oops , I see it now, good catch ;)

**Pull request merged**